test/blackbox/run_blackbox_tests/test_cases_set_metaconn_retry_cb.c

```diff
@@ -0,0 +1,130 @@
+    with this program; if not, write to the Free Software Foundation, Inc.,
+    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#include "execute_tests.h"
+#include "test_cases_set_metaconn_retry_cb.h"
+#include "../common/test_step.h"
+#include "../common/common_handlers.h"
+#include "../../utils.h"
+#include <assert.h>
+#include <setjmp.h>
+#include <cmocka.h>
+#include <pthread.h>
+
+static void test_case_set_mataconn_retry_cb_01(void **state);
+static bool test_set_mataconn_retry_cb_01(void);
+
+static bool status;
+static int retry_attempts;
+static struct sync_flag join_cond = {.mutex  = PTHREAD_MUTEX_INITIALIZER, .cond = PTHREAD_COND_INITIALIZER};
+
+static void status_cb(meshlink_handle_t *mesh, meshlink_node_t *source, bool reach) {
+	if(!strcmp(mesh->name, "foo") && !strcmp(source->name, "bar")) {
+		status = reach;
+		set_sync_flag(&join_cond, true);
+	}
+}
+
+/* Meta-connection retry callback handler */
+static void meta_retry_cb(meshlink_handle_t *mesh, meshlink_node_t *source) {
+	if(!strcmp(mesh->name, "foo")) {
+		++retry_attempts;
+	}
+}
+
+/* Execute set meta connection retry callback Test Case # 1 */
+static void test_case_set_mataconn_retry_cb_01(void **state) {
+	execute_test(test_set_mataconn_retry_cb_01, state);
+}
+
+/* Test Steps for meshlink_set_metaconn_retry_cb Test Case # 1
+
+    Test Steps:
+
+    Expected Result:
+*/
+static bool test_set_mataconn_retry_cb_01(void) {
+	meshlink_destroy("meshlink_conf.1");
+	meshlink_destroy("meshlink_conf.2");
+
+	// Opening foo and bar nodes
+	meshlink_handle_t *mesh1 = meshlink_open("meshlink_conf.1", "foo", "test", DEV_CLASS_STATIONARY);
+	assert(mesh1 != NULL);
+	meshlink_set_log_cb(mesh1, MESHLINK_DEBUG, meshlink_callback_logger);
+	meshlink_enable_discovery(mesh1, false);
+	meshlink_handle_t *mesh2 = meshlink_open("meshlink_conf.2", "bar", "test", DEV_CLASS_STATIONARY);
+	assert(mesh2 != NULL);
+
+	// Set up callback for node status
+	meshlink_set_node_status_cb(mesh1, status_cb);
+	meshlink_set_connection_try_cb(mesh1, meta_retry_cb);
+
+	// Exporting and Importing mutually
+	char *exp1 = meshlink_export(mesh1);
+	assert(exp1 != NULL);
+	char *exp2 = meshlink_export(mesh2);
+	assert(exp2 != NULL);
+	bool imp1 = meshlink_import(mesh1, exp2);
+	bool imp2 = meshlink_import(mesh2, exp1);
+	free(exp1);
+	free(exp2);
+
+	assert(meshlink_start(mesh1));
+	assert(meshlink_start(mesh2));
+
+	// Wait for foo and bar nodes to join
+	assert(wait_sync_flag(&join_cond, 5));
+	assert(status);
+
+	// Joining should in this case raise one connection try callback
+	assert_int_equal(retry_attempts, 1);
+
+	// Close the bar node
+	set_sync_flag(&join_cond, false);
+	meshlink_close(mesh2);
+	assert(wait_sync_flag(&join_cond, 5));
+	assert(!status);
+
+	// Sleep for 31 seconds and it should raise 3 connection try callbacks after closing bar node.
+	sleep(31);
+	meshlink_close(mesh1);
+	assert_int_equal(retry_attempts, 4);
```

Comment:
This looks a bit fragile to me. It checks whether if we sleep for approximately 31 seconds in this thread, that the MeshLink thread will have made exactly 3 attempts to reconnect. But timers are not always very precise, and we might want to change the interval between retry attempts in the future. Maybe it is better to just wait for a given number of reconnection attempts to have occurred, for example by using another sync flag that is set by the connection_try callback.